### ISSUE TITLE
Performance breakage of dslash

### ISSUE THREAD

**Issue by paboyle** [paboyle/Grid#165]
I’ve been unable to reproduce the ITT benchmark performance on our new machine
with the current Grid develop. 

After (four) days of tracking this, it appears Chris Kelly's commits of the
new G-Parity hand unrolled kernels halved the performance of the original code, and leaves them slower than the compiler generated “generic” kernels under Intel icpc 2018.

git checkout 59bd1fe21b3fc81b8c50e9caa53f2adc65b3d7b5 WilsonKernelsHand.cc   (end of C Kelly edits)
Overlapped comms/compute for 8^4,12^4, 16^4, 24^4
Wilson
Grid : Message : Average mflops/s per call per node (full): 7829.6
Grid : Message : Average mflops/s per call per node (full): 28028.6
Grid : Message : Average mflops/s per call per node (full): 64968.1
Grid : Message : Average mflops/s per call per node (full): 96672.8
DWF
Grid : Message : Average mflops/s per call per node (full): 51220.2
Grid : Message : Average mflops/s per call per node (full): 95553.3
Grid : Message : Average mflops/s per call per node (full): 99008.4
Grid : Message : Average mflops/s per call per node (full): 107145.4
sDWF
Grid : Message : Average mflops/s per call per node (full): 54373.1
Grid : Message : Average mflops/s per call per node (full): 95631.2
Grid : Message : Average mflops/s per call per node (full): 95689.5
Grid : Message : Average mflops/s per call per node (full): 100096.6

git checkout f301be94ce9e0c08a12482b69dc131fa82fd2a0d WilsonKernelsHand.cc
Wilson
Grid : Message : Average mflops/s per call per node (full): 7997.5
Grid : Message : Average mflops/s per call per node (full): 28937.1
Grid : Message : Average mflops/s per call per node (full): 66775.1
Grid : Message : Average mflops/s per call per node (full): 165339.6
DWF
Grid : Message : Average mflops/s per call per node (full): 52525.5
Grid : Message : Average mflops/s per call per node (full): 145692.8
Grid : Message : Average mflops/s per call per node (full): 184362.0
Grid : Message : Average mflops/s per call per node (full): 202913.3
sDWF
Grid : Message : Average mflops/s per call per node (full): 55750.0
Grid : Message : Average mflops/s per call per node (full): 163095.1
Grid : Message : Average mflops/s per call per node (full): 184847.4
Grid : Message : Average mflops/s per call per node (full): 185433.7

The Parity kernels should have been written separately, and left the original, normal BC Wilson kernels unmodified.  Changing the original was a hacky and risky approach to take.
There were clear template specialisation hooks left empty for Gparity to fill, and these should have been used without touching the original code & macros which were developed very carefully.

I will revert the code to the original fast version, and implement the G-Parity template specialisation
in a separate file, (WilsonKernelsHandGparity.cc) with your updated code only instantiated there. 

However, I’m puzzled if there is the proper benefit from the hand unrolled Gparity if it destroys the 
performance of the original. There is probably a lot of potential optimisation missing.

Adding new dirac operators is benign, but modifying existing code that has been painstakingly developed is almost never performance neutral and inadequate care was used.

Bottom line: DO NOT TOUCH THE CODE ENTERING THE CORE DIRAC OPERATOR ROUTINES.

**Comment by giltirn** [comment]
Dear Peter,

I'm very sorry that my change caused you so much hassle. Although it was 
a while ago, I do recall the changes being primarily in passing macro 
variable names that were formerly baked in. I was careful to test that 
my changes did not change the performance of the non-Gparity kernel on 
KNL and that the performance of the Gparity version equaled if not 
exceeded that of the non-Gparity version. Presumably this means that the 
performance issues you are seeing are only on the Skylake systems, or 
due to compiler changes. Do you have any insight into what changed in 
the preprocessed code?

Best,
Chris

**Comment by paboyle** [comment]
Hi Chris,

not yet; I split into two files one with "before" and one with "after" and instantiated the Gparity
only in the "after" version. I think we should revisit the gparity later, because if it is generating slower wilson code, then presumably this indicates that the audit you propose might turn up further optimisations.

The difference so far is on:

   Skylake silver, under AVX2 with Intel 2018 compiler

Peter

**Comment by paboyle** [comment]
this was fixed long ago. Closing. Didnt realise it was still open
